### DISCUSSION TITLE
Grammar Import Loop/Race Condition?

### DISCUSSION THREAD

**Discussion by snarkipus** [eclipse-langium/langium#879] [Q&A] [answered]
Tags: @msujew 

So, given the obscene size of the grammar I'm working with, I refactored it to take advantage of the `import` capability. 

```terminal
Attempted reference resolution before document reached ComputedScopes state (file:///c%3A/Users/Me/Projects/name/examples/project/src/language-server/grammar-partA.langium).
Attempted reference resolution before document reached ComputedScopes state (file:///c%3A/Users/Me/Projects/name/examples/project/src/language-server/grammar-partA.langium).
Attempted reference resolution before document reached ComputedScopes state (file:///c%3A/Users/Me/Projects/name/examples/project/src/language-server/grammar-partA.langium).
...
[17:03:45] src\language-server\grammar-top.langium:22:5 - Could not resolve reference to AbstractRule named 'AbstractRuleA'.
[17:03:45] src\language-server\grammar-top.langium:22:22 - Could not resolve reference to AbstractRule named 'AbstractRuleB'.
[17:03:45] src\language-server\grammar-top.langium:22:39 - Could not resolve reference to AbstractRule named 'AbstractRuleC'.
[17:03:45] Langium generator failed.
```
The easiest way for me to explain the import structure is to illustrate: 

![langium import loop](https://user-images.githubusercontent.com/3688989/213818599-45e00ad2-f273-4720-bf26-c4624ad6d947.png)

The import in  $\color{purple}\textsf{purple}$ is the one that breaks it. I'm trying to recreate a example case but haven't had any luck yet. I'll update it here if I can scale it down ... but the "working" implementation is ~~100% repeatable~~ _weirdly_ repeatable.

Thoughts?

### Update
Two points of interest:
 - per the diagram, `langium:watch` only triggers on changes to `grammar-top.langium` ... is this expected?
 - probably important - the `generate` fails once, then passes on the subsequent run (i.e. first run fails, second run passes)

**Comment by msujew** [comment, marked as answer]
> Thoughts?

Likely a bug somewhere in our generator. Probably related to copying rules into a "super grammar" which contains all used rules.

> per the diagram, langium:watch only triggers on changes to grammar-top.langium ... is this expected?

Kind of, yes. It currently only listens to changes in the file that's referenced in the langium config. We should probably propagate any referenced files back to the file listener.

> probably important - the generate fails once, then passes on the subsequent run (i.e. first run fails, second run passes)

Interesting. There's likely something weird going on in our generator, which tries to resolve references before the content has been processed as indicated by the warning.

**Reply by spoenemann** [reply]
I suggest creating two issues from this discussion, one for fixing the generator and another for improving watch mode to include referenced grammars.

**Reply by snarkipus** [reply]
Created #882 with a minimal repo to reproduce.
